File: Competive Programming/C++/HackerRank/marcs-cakewalk.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>
using namespace std;


long marcsCakewalk(vector<int> calories, int n) {
    long sum = 0;
    sort(calories.rbegin(), calories.rend());

    for (int i=0; i < n; i++)
        sum += ((unsigned long long)pow(2, i) * calories.at(i));
    

    return sum;
}


int main() {
    int n, tmp;
    cin >> n;

    vector<int> calories;
    for (int i=0; i < n; ++i) {
        cin >> tmp; 
        calories.push_back(tmp);
    }

    cout << marcsCakewalk(calories, n) << endl;
    return 0;
}

File: Competive Programming/Go/HackerRank/marcs-cakewalk.go
package main

import (
    "fmt"
    "sort"
    "math"
)

func marcsCakewalk(calories []int, len int) int {
    // The comparator function takes indexes not actual values
    sort.Slice(calories, func(i, j int) bool { return calories[i] > calories[j] })
    var sum int = 0

    for k, v := range calories {
        sum += (int(math.Pow(2.0, float64(k))) * v)
    }

    return sum;
}

func main() {
    var n int
    fmt.Scanf("%d", &n);

    calories := make([]int, n)
    for i := 0; i < n; i++ {
        fmt.Scanf("%d", &calories[i])
    }

    fmt.Println(marcsCakewalk(calories, n))
}
